includes/cli/class-data-backfill.php

```diff
@@ -0,0 +1,365 @@
+					'relation' => 'AND',
+					[
+						'key'     => 'timestamp',
+						'value'   => $timestamp,
+						'compare' => '=',
+					],
+					[
+						'key'     => 'action_name',
+						'value'   => $action,
+						'compare' => '=',
+					],
+					[
+						'key'     => 'data',
+						'value'   => wp_json_encode( $data ),
+						'compare' => '=',
+					],
+				],
+			]
+		);
+	}
+
+	/**
+	 * Process reader_registered events.
+	 *
+	 * @param string $start The start date.
+	 * @param string $end The end date.
+	 * @param bool   $live Whether to run in live mode.
+	 * @param bool   $verbose Whether to output verbose information.
+	 */
+	private static function process_reader_registered( $start, $end, $live, $verbose ) {
+		// Get all users registered between $start and $end.
+		$users = get_users(
+			[
+				'role__in'   => \Newspack\Reader_Activation::get_reader_roles(),
+				'date_query' => [
+					'after'     => $start,
+					'before'    => $end,
+					'inclusive' => true,
+				],
+				'fields'     => [ 'id', 'user_email', 'user_registered' ],
+				'number'     => -1,
+				'meta_query' => [ // phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_query
+					[
+						'key'     => \Newspack_Network\Utils\Users::USER_META_REMOTE_SITE,
+						'compare' => 'NOT EXISTS',
+					],
+				],
+			]
+		);
+		if ( ! $verbose ) {
+			self::$progress = \WP_CLI\Utils\make_progress_bar( 'Processing users', count( $users ) );
+		}
+
+		foreach ( $users as $user ) {
+			$registration_method = get_user_meta( $user->ID, \Newspack\Reader_Activation::REGISTRATION_METHOD, true );
+			$user_data = [
+				'user_id'  => $user->ID,
+				'email'    => $user->user_email,
+				'metadata' => [
+					// 'current_page_url' is not saved, can't be backfilled.
+					'registration_method' => empty( $registration_method ) ? 'backfill-script' : $registration_method,
+				],
+			];
+			if ( $live ) {
+				self::process_event_entity( $user_data, strtotime( $user->user_registered ), 'reader_registered' );
+			}
+			if ( $verbose ) {
+				WP_CLI::line( '👉 ' . sprintf( 'User %s (#%d) registered on %s.', $user->user_email, $user->ID, $user->user_registered ) );
+			}
+			if ( self::$progress ) {
+				self::$progress->tick();
+			}
+		}
+	}
+
+	/**
+	 * Process donation_new events.
+	 *
+	 * @param string $start The start date.
+	 * @param string $end The end date.
+	 * @param bool   $live Whether to run in live mode.
+	 * @param bool   $verbose Whether to output verbose information.
+	 */
+	private static function process_donation_new( $start, $end, $live, $verbose ) {
+		$orders = wc_get_orders(
+			[
+				'status'       => 'completed',
+				'date_created' => $start . '...' . $end,
+				'limit'        => -1,
+			]
+		);
+		if ( ! $verbose ) {
+			self::$progress = \WP_CLI\Utils\make_progress_bar( 'Processing orders', count( $orders ) );
+		}
+		foreach ( $orders as $order ) {
+			$order_id = $order->get_id();
+			$order_data = \Newspack\Data_Events\Utils::get_order_data( $order_id );
```

Comment:
we need to check if it is a donation related order. see my review on the newspack plugin's PR

Reply:
Addressed in the `newspack-plugin` PR